### PR TITLE
A big hecking chemistry-related refactor.

### PR DESCRIPTION
**Pull request by PJB3005**
Changed `SolutionContainerCaps`. It now describes "stock" behavior for interacting with solutions that is pre-implemented by `SolutionContainerComponent`. As such things like syringes do not check it anymore (on themselves) to see "can we remove reagent from ourselves". That's assumed by it... being a syringe.

`SolutionContainerCaps` now has different flags more accurately describing possible reagent interaction behaviors.

`ISolutionInteractionsComponent` is the interface that describes the common behaviors like "what happens when injected with a syringe". This is implemented by `SolutionContainerComponent` but could be implemented by other classes. One notable example that drove me to making this interface was the /vg/station circuit imprinter which splits reagent poured in into its two reservoir beakers. Having this interface allows us to do this "proxying" behavior hack-free. (the hacks in /vg/ code were somewhat dirty...).

This was the primary catalyst for this PR: You can't remove from other syringes with syringes anymore.

`PourableComponent` has been replaced `SolutionTransferComponent`. It now describes both give-and-take behavior for the common reagent containers. This is in line with /vg/'s `/obj/item/weapon/reagent_container` architecture. "Taking" in this context is ONLY from reagent tanks like fuel tanks.

Oh, should I mention that fuel tanks and such have a proper component now? They do.

Because of this behavioral change, reagent tanks DO NOT have `Pourable` anymore. Removing from reagent tanks is now in the hands of the item used on them. Welders and fire extinguishers now have code for removing from them. This sounds bad at first but remember that all have quite unique behavior related to this: Welders cause explosions if lit and can ONLY be fueled at fuel tanks. Extinguishers can be filled at any tank, etc... The code for this is also simpler due to `ISolutionInteractionsComponent` now so...

`IAfterInteract` now works like `IInteractUsing` with the `Priority` levels and "return true to block further handlers" behavior. This was necessary to make extinguishers prioritize taking from tanks over spraying.

Explicitly coded interactions like welders refueling also means they refuse instantly to full now, which they didn't before. And it plays the sound. Etc...

Probably more stuff I'm forgetting.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(this, c => c.TransferAmount, "TransferAmount", ReagentUnit.New(10));
+            serializer.DataField(this, c => c.TankType, "TankType", ReagentTankType.Unspecified);
+        }
+    }
+
+    public enum ReagentTankType
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs:30]:
```suggestion
    public enum ReagentTankType : byte
```

File: Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    /*
+    - type: WeldingTankComponent
+      explosionStrength:
+        dev: 1
+        heavy: 3
+        light: 5
+        flash: 7
+     */
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs:51]: Is this even used?

File: Content.Shared/Chemistry/SolutionContainerCaps.cs [hunk 3]
```diff
@@ -1,39 +1,53 @@
...
+
+        /// <summary>
+        ///     Reagents can be easily added via all reagent containers.
+        /// </summary>
+        Refillable = 1 << 2,
+
+        /// <summary>
+        ///     Reagents can be easily removed through any reagent container.
+        /// </summary>
+        Drainable = 1 << 3,
```
Comment (metalgearsloth) [inline Content.Shared/Chemistry/SolutionContainerCaps.cs:35]: I think I get what these are but maybe have an example? e.g. beakers, cans?

**Review by metalgearsloth — COMMENTED** [review]
Seems good, also credit refill.ogg if not already done.

One thing I noticed that's more of a drinks issue is that it's possible to use the "Spill liquid" verb + draw from a drink can even when it hasn't been popped (I guess injecting makes sense intuitively but the spilling doesn't?)

File: Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public ReagentTankType TankType { get; set; }
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+
+            serializer.DataField(this, c => c.TransferAmount, "TransferAmount", ReagentUnit.New(10));
+            serializer.DataField(this, c => c.TankType, "TankType", ReagentTankType.Unspecified);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs:26]: Should be transferAmount and tankType

File: Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs (hunk 2 above)
Comment (PJB3005) [reply Content.Server/GameObjects/Components/Chemistry/ReagentTankComponent.cs:51]: Oh I typed this out while giving an explanation on Discord. My bad.

File: Content.Server/GameObjects/Components/Chemistry/SolutionTransferVerbs.cs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        public override void GetData(IEntity user, IEntity target, VerbData data)
+        {
+            if (!target.TryGetComponent(out ISolutionInteractionsComponent? targetSolution) ||
+                !ActionBlockerSystem.CanInteract(user) ||
+                !user.TryGetComponent<HandsComponent>(out var hands) ||
+                hands.GetActiveHand == null ||
+                hands.GetActiveHand.Owner == target ||
+                !hands.GetActiveHand.Owner.TryGetComponent(out ISolutionInteractionsComponent? sourceSolution) ||
+                !sourceSolution.CanDrain ||
+                !targetSolution.CanRefill)
```
[the remark this answers is not in the document]
Comment (PJB3005) [reply Content.Server/GameObjects/Components/Chemistry/SolutionTransferVerbs.cs]: I made it somewhat less bad but I doubt I can compress it more.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**